source/http_server.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include "http_server.h"
+
+#include "io.h"
+
+#include <aws/common/array_list.h>
+#include <aws/http/request_response.h>
+#include <aws/http/server.h>
+#include <aws/io/socket.h>
+#include <aws/io/stream.h>
+
+#ifdef _WIN32
+#    define LOCAL_SOCK_TEST_FORMAT "\\\\.\\pipe\\testsock-%s"
+#else
+#    define LOCAL_SOCK_TEST_FORMAT "testsock-%s.sock"
+#endif
+
+const char *s_capsule_name_http_server = "aws_http_server";
+
+struct py_http_server {
+    struct aws_allocator *allocator;
+    struct aws_http_server *server;
+    PyObject *capsule;
+    PyObject *on_incoming_connection;
+    PyObject *on_destroy_complete;
+    bool destructor_called;
+    bool destroy_called;
+    bool destroy_complete;
+};
+
+static void s_http_server_destructor(PyObject *http_server_capsule) {
+    struct py_http_server *py_server = PyCapsule_GetPointer(http_server_capsule, s_capsule_name_http_server);
+    assert(py_server);
+    py_server->destructor_called = true;
+    if (py_server->server) {
+        if (!py_server->destroy_called) {
+            aws_http_server_release(py_server->server);
+            py_server->destroy_called = true;
+        }
+        py_server->server = NULL;
+    }
+    /* the incoming callback is not freed until now */
+    Py_DECREF(py_server->on_incoming_connection);
+    if (py_server->destroy_complete) {
+        aws_mem_release(py_server->allocator, py_server);
+    }
+}
+
+static void s_on_destroy_complete(void *user_data) {
+    struct py_http_server *py_server = user_data;
+    PyGILState_STATE state = PyGILState_Ensure();
+    py_server->destroy_complete = true;
+    PyObject *on_destroy_complete_cb = py_server->on_destroy_complete;
+    PyObject *result = NULL;
+    if (!py_server->destructor_called && on_destroy_complete_cb) {
+        result = PyObject_CallFunction(on_destroy_complete_cb, "(N)", py_server->capsule);
+        Py_DECREF(on_destroy_complete_cb);
```

Comment:
```
    if (result) {
        Py_DECREF(result);
    } else {
        PyErr_WriteUnraisable(PyErr_Occurred());
    }
```

Some things about [PyObject_CallFunction()](https://docs.python.org/3/c-api/object.html#c.PyObject_CallFunction):
1) If `result` is non-null, you need to decref it
2) If `result` IS null then an exception has occurred. Either the callback couldn't be called, or an exception was raised within the callback that didn't get caught. There's not much we can do now that we're back in C-Town, but we can use [PyErr_WriteUnraisable()](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WriteUnraisable) to print a msg to stderr where the user might see it